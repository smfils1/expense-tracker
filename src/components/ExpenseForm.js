import React from 'react';
import moment from 'moment'
import {SingleDatePicker} from 'react-dates';


export default class ExpenseForm extends React.Component{
    state = {
        description: this.props.expense ? this.props.expense.description : '',
        amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
        note: this.props.expense ? this.props.expense.note : '',
        createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
        calendarFocused: false,
        error: ''
    }
    
    render(){
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    {this.state.error &&<p>{this.state.error}</p> }

                    <input type="text" 
                    placeholder="Description"
                    name="description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}/>
                    <input type="text" placeholder="Amount"
                    name="amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}/>
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={()=> false} //Make all past days available
          />
                    <textarea name="note" placeholder="Add a note (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}></textarea>
                    <button>Add Expense</button>
                    
                </form>
                
            </div>
        )
    }
    onDescriptionChange = (e)=>{
        const description = e.target.value
        this.setState(()=>({
            description
        }))
    }
    onAmountChange = (e)=>{
        const amount = e.target.value
        if(amount.match(/^\d*(\.\d{0,2})?$/)){
            this.setState(()=>({
                amount
            }))            
        }

    }
    onNoteChange = (e)=>{
        const note = e.target.value
        this.setState(()=>({
            note
        }))
    }

    onDateChange = (createdAt)=>{
        this.setState(()=>({
            createdAt
        }))
    }
    onFocusChange = ({ focused }) => {
        this.setState({ calendarFocused:focused })
    }
    
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
          this.setState(()=>({
            error: 'Please enter a description and/or amount'
          }))
        } else{
          this.setState(()=>({
            error: ''
          }))
          this.props.onSubmit({
            description: this.state.description,
            amount:parseFloat(this.state.amount,10)*100,
            note:this.state.note,
            createdAt: this.state.createdAt.valueOf(),
          })

        }
    }

}