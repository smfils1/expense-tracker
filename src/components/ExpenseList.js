import React from "react";
import { connect } from "react-redux";
import ListItem from "./ExpenseListItem";
import {getVisibleExpenses} from '../selectors/expenses'

export const ExpenseList = ({expenses}) => (
  <div>
    <h1>Expenses</h1>
    {
      expenses.length === 0 ? <p>No Expenses</p> :
      expenses.map(item => (
      <ListItem {...item} key={item.id} />   
    ))
    }    
  </div>
);

const mapStateToProps = state => ({
  expenses: getVisibleExpenses(state.expenses, state.filter)
});

export default connect(mapStateToProps)(ExpenseList);
