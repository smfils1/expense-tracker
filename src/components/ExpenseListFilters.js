import React from "react";
import { connect } from "react-redux";
import uuid from 'uuid'
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from "../actions/filters";


import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  }

  onDateChange = ({startDate,endDate}) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }
  onFocusChange = (focused) => {
    this.setState(()=>({ calendarFocused: focused }));
  }

  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  }
  onSortByChange = e => {
    const filter = e.target.value;
    if (filter === "date") {
      return this.props.sortByDate();
    }
    this.props.sortByAmount();
  }


  render() {
    return (
      <div>
        <input
          name="text"
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select
          name="sortBy"
          value={this.props.filters.sortBy}
          onChange={this.onSortByChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId={uuid()}
          endDateId ={uuid()}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDateChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={()=>false}
          showClearDates={true}
        


        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filter
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate:() => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
  

});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
