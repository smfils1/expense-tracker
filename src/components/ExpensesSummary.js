import React from 'react';

import expensesTotal from '../selectors/expensesTotal'
import {connect} from 'react-redux'
import {getVisibleExpenses} from '../selectors/expenses'

export const ExpensesSummary = ({expensesCount=0,expensesTotalAmount=0}) => (
    <div>
        <h3>Viewing {expensesCount} expenses totalling ${expensesTotalAmount}</h3>
    </div>
)

const mapStateToProps = (state) =>{
    const expenses =getVisibleExpenses(state.expenses, state.filter)
    return {
        expensesCount: expenses.length,
        expensesTotalAmount: expensesTotal(expenses) / 100
    }
}

export default connect(mapStateToProps)(ExpensesSummary)

