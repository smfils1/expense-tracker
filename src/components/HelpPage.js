import React from 'react';

const HelpPage = () => (
    <p>Help Content</p>   
)

export default HelpPage