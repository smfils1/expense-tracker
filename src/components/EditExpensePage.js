import React from 'react';
import ExpenseForm from'./ExpenseForm'
import { connect } from "react-redux";
import {editExpense, removeExpense} from '../actions/expenses'



export class EditExpensePage extends React.Component{
    onClick = () =>{
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push('/')     

    }    
    onSubmit = expense =>{
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push('/')     

    }
    render(){
        return (
            <div>
                <p>Edit Content: {this.props.match.params.id}</p>
                <ExpenseForm
                expense={this.props.expense} 
                onSubmit={this.onSubmit}/>
                
                <button onClick={this.onClick}>Remove</button>
            </div>

)
    }
} 


const mapStateToProps = (state,props)=> ({
    expense: state.expenses.find((expense)=>expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch)=> ({
    editExpense:(id, expense)=>{
        dispatch(editExpense(id, expense)) 
    },
    removeExpense:(id)=>{
        dispatch(removeExpense(id))
    }

})

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);
