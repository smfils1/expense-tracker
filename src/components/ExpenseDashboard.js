import React from 'react';
import ExpenseList from './ExpenseList';
import ExpensesSummary from './ExpensesSummary';

import ExpenseListFilters from './ExpenseListFilters'


const ExpenseDashboard = (props) => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilters/>
        <ExpenseList/>         
    </div>

)


export default ExpenseDashboard