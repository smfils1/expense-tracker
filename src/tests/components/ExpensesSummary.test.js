import React from 'react'
import {shallow} from 'enzyme'
import {ExpensesSummary} from '../../components/ExpensesSummary'


describe('<ExpensesSummary>',()=>{

    test('Should correctly render ExpenseSummary w/ zero expense by default',()=>{
        const wrapper = shallow(<ExpensesSummary/>)
        expect(wrapper).toMatchSnapshot()
    })
    test('Should correctly render ExpenseSummary w/ one expense',()=>{
        const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotalAmount={100}/>)
        expect(wrapper).toMatchSnapshot()
    })
    test('Should correctly render ExpenseSummary w/ multiple expenses',()=>{
        const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotalAmount={100}/>)
        expect(wrapper).toMatchSnapshot()
    })
})