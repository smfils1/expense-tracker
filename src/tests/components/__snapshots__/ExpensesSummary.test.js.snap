// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpensesSummary> Should correctly render ExpenseSummary w/ multiple expenses 1`] = `
<div>
  <h3>
    Viewing 
    2
     expenses totalling $
    100
  </h3>
</div>
`;

exports[`<ExpensesSummary> Should correctly render ExpenseSummary w/ one expense 1`] = `
<div>
  <h3>
    Viewing 
    1
     expenses totalling $
    100
  </h3>
</div>
`;

exports[`<ExpensesSummary> Should correctly render ExpenseSummary w/ zero expense by default 1`] = `
<div>
  <h3>
    Viewing 
    0
     expenses totalling $
    0
  </h3>
</div>
`;
